Data/Graph/Inductive/Query/Cycles.hs

```diff
@@ -0,0 +1,178 @@
+{- |
+   Module      : Data.Graph.Inductive.Query.Cycles
+   Description : Finds all cycles.
+   Copyright   : (c) Gabriel Hjort Blindell <gabriel.hjort.blindell@gmail.com>
+                     Ivan Lazar Miljenovic <Ivan.Miljenovic@gmail.com>
+   License     : BSD3
+-}
+
+module Data.Graph.Inductive.Query.Cycles
+  ( cycles
+  , cycles'
+  , uniqueCycles
+  , uniqueCycles'
+  )
+where
+
+import Data.Graph.Inductive.Graph
+import Data.Graph.Inductive.Query.SCC
+
+import Data.List ((\\), delete, tails)
+import Data.Maybe (fromJust)
+import Control.Monad (ap)
+import qualified Data.IntMap as M
+
+
+
+-- The following functions were copied from the Graphalyze package.
+
+-- | Obtain the labels for a list of 'Node's. It is assumed that each 'Node' is
+-- indeed present in the given graph.
+addLabels :: (Graph g) => g a b -> [Node] -> [LNode a]
+addLabels gr = map (ap (,) (fromJust . lab gr))
+
+twoCycle      :: (Graph g) => g a b -> Node -> [Node]
+twoCycle gr n = filter (elem n . suc gr) (delete n $ suc gr n)
+
+-- | Determines if the list of nodes represents a regular subgraph.
+isRegular :: (Graph g) => g a b -> [Node] -> Bool
+isRegular g ns = all allTwoCycle split
+    where
+      -- Node + Rest of list
+      split = zip ns tns'
+      tns' = tail $ tails ns
+      allTwoCycle (n,rs) = null $ rs \\ twoCycle g n
+
+-- End of copied functions.
+
+
+
+-- | Contains the necessary data structures used by 'cycles'.
+data CyclesInState g a b
+  = CyclesInState
+      { cisCycles :: [[Node]]
+        -- ^ The cycles found so far, in topological order.
+      , cisBlocked :: M.IntMap Bool
+        -- ^ The nodes which are currently blocked.
+      , cisBlockMap :: M.IntMap [Node]
+        -- ^ The B set.
+      , cisStack :: [Node]
+        -- ^ The node stack.
+      , cisS :: Maybe Node
+        -- ^ The current S value.
+      , cisCurrentComp :: Maybe (g a b)
+        -- ^ The component currently being processed.
+      , cisComponents :: [g a b]
+        -- ^ The components of the input graph.
+      , cisGraph :: g a b
+        -- ^ The input graph.
+      }
+  deriving (Show, Read, Eq)
+
+-- | Finds all cycles in a given graph. The returned lists contains the nodes
+-- appearing in the cycles, in successor order (although it is undefined which
+-- node appears first).
+--
+-- Implemented using Johnson's algorithm. See Donald B. Johnson: Finding All the
+-- Elementary Circuits of a Directed Graph. SIAM Journal on Computing. Volumne
+-- 4, Nr. 1 (1975), pp. 77-84.
+cycles :: (Graph g) => g a b -> [[LNode a]]
+cycles g = map (addLabels g) (cycles' g)
+
+-- | Same as 'cycles' but for unlabeled graphs.
```

Comment:
Not for unlabelled graphs, just doesn't return the labels.

Reply:
Fixed.